src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -154,48 +159,54 @@ protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
 
             return HttpContext.Response.Body.FlushAsync();
         }
-		
+
         public void Initialize()
         {
             var timeout = GetTimeout();
 
             if (timeout != TimeSpan.Zero)
             {
-                // CancellationTokenSource does not support greater than int.MaxValue milliseconds
-                if (timeout.TotalMilliseconds > int.MaxValue)
-                {
-                    throw new InvalidOperationException("A timeout greater than 2147483647 milliseconds is not supported.");
-                }
-
                 _deadline = Clock.UtcNow.Add(timeout);
-                _cts = new CancellationTokenSource(timeout);
+
+                _deadlineTimer = new Timer(DeadlineExceeded, timeout, timeout, Timeout.InfiniteTimeSpan);
```

Comment:
This is a common enough pattern (abort the connection after some timeout, that I think we should make it a first class API).